### PR TITLE
Sarkars/flib ngenc compute

### PR DESCRIPTION
**Pull request by sayantan-nervana**

### CODE REVIEW

**Comment by sayantan-nervana** [comment]
GPU TF python errors:

```
math_ops_test.AccumulateNTest.testInt ..ERROR
--
  | math_ops_test.AccumulateNTest.testFloat ..ERROR
  | reduction_ops_test.AllReductionTest.testAxesType ..ERROR
  | reduction_ops_test.AnyReductionTest.testAxesType ..ERROR
  | cast_op_test.CastOpTest.testInfNan ..FAIL
  | math_ops_test.DivAndModTest.testRealDiv ..FAIL
  | reduction_ops_test.SumReductionTest.testHighRank ..FAIL
  | relu_op_test.ReluTest.testGradientFloat64 ..FAIL


```

File: tools/test_utils.py [hunk 1]
```diff
@@ -46,9 +47,8 @@ def install_ngraph_bridge(artifacts_dir):
             print("Existing Wheel: " + whl)
         raise Exception("Error getting the ngraph-tf wheel file")
 
-    # First ensure that we have nGraph installed
     ng_whl = os.path.join(artifacts_dir, ngtf_wheel_files[0])
-    command_executor(["pip", "install", "-U", ng_whl])
+    subprocess.call([sys.executable, "-m", "pip", "install", ng_whl])
```
Comment (sayantan-nervana) [inline tools/test_utils.py]: TODO: remove once https://github.com/tensorflow/ngraph-bridge/tree/sarkars/current_process_py_to_pip_install is merged

File: src/ngraph_utils.cc [hunk 2]
```diff
@@ -378,7 +378,18 @@ void AllreduceOpControlOrder(
       }
     }
   }
-};
+}
+
+bool IsProcessedByNgraphPass(Graph* g) {
```
Comment (sayantan-nervana) [inline src/ngraph_utils.cc]: This addition is the main change wrt the old PR.

Please note where its called to disable ngraphs rewrite pass

File: src/enable_variable_ops/ngraph_utils.cc [hunk 3]
```diff
@@ -324,6 +324,18 @@ void MemoryProfile(long& vm_usage, long& resident_set) {
   }
 };
 
+// TODO: Merge with normal path utils
+bool IsProcessedByNgraphPass(Graph* g) {
+  // TODO: place a dummy node as a marker
+  // Current method may fail when graph had no encapsulates or ngvars after
```
Comment (kanvi-nervana) [inline src/enable_variable_ops/ngraph_utils.cc]: Minor:  'has' instead of 'had'

File: src/ngraph_mark_for_clustering.cc [hunk 4]
```diff
@@ -291,6 +291,7 @@ Status MarkForClustering(Graph* graph,
       confirmation_function_map["Minimum"] = SimpleConfirmationFunction();
       confirmation_function_map["Mul"] = SimpleConfirmationFunction();
       confirmation_function_map["Neg"] = SimpleConfirmationFunction();
+      confirmation_function_map["NoOp"] = SimpleConfirmationFunction();
```
Comment (kanvi-nervana) [inline src/ngraph_mark_for_clustering.cc]: Why do we need this?

File: src/ngraph_rewrite_pass.cc [hunk 5]
```diff
@@ -123,7 +124,19 @@ class NGraphVariableCapturePass : public NGraphRewritePass {
...
+      // populated. Then we run a new network, it repopulates the cluster
+      // manager. This works under the assumption that whenever
+      // NGraphEncapsulate's Compute is run the rewrite passes (grappler or
+      // optimization passes) have also run (compute --> rewrite). Now that
+      // assumption is broken because now we support NGraphEncapsulate enabled
+      // graphs. Such graphs will not run the first rewrite pass, hence the
+      // cluster manager is not overwritten. Which would mean that cluster
+      // manager contains stale data from a previous run. Hence evicting cluster
+      // manager when rewrite passes are not run.
+      NGraphClusterManager::EvictAllClusters();
```
Comment (sindhu-nervana) [inline src/ngraph_rewrite_pass.cc]: Thanks for adding this comment/explanation !

File: src/enable_variable_ops/ngraph_rewrite_pass.cc [hunk 6]
```diff
@@ -163,8 +165,14 @@ class NGraphVariableCapturePass : public NGraphRewritePass {
     // If ngraph is disabled via ngraph_bridge api or NGRAPH_TF_DISABLE is set
     // we will not do anything; all subsequent
     // passes become a no-op.
-    if (config::IsEnabled() == false ||
-        std::getenv("NGRAPH_TF_DISABLE") != nullptr) {
+    bool ngraph_not_enabled =
+        (!config::IsEnabled()) || (std::getenv("NGRAPH_TF_DISABLE") != nullptr);
+    bool already_processed = IsProcessedByNgraphPass(options.graph->get());
+    if (ngraph_not_enabled || already_processed) {
+      NGRAPH_VLOG(0) << "Not running through nGraph. nGraph not enabled: "
```
Comment (avijit-nervana) [inline src/enable_variable_ops/ngraph_rewrite_pass.cc]: We don't want VLOG(0).

File: src/enable_variable_ops/ngraph_rewrite_pass.cc [hunk 7]
```diff
@@ -232,8 +240,14 @@ class NGraphEncapsulationPass : public NGraphRewritePass {
...
```
Comment (avijit-nervana) [inline src/enable_variable_ops/ngraph_rewrite_pass.cc]: VLOG(0)?

File: src/grappler/ngraph_optimizer.cc [hunk 8]
```diff
@@ -63,9 +64,14 @@ Status NgraphOptimizer::Optimize(tensorflow::grappler::Cluster* cluster,
   // If ngraph is disabled via ngraph_bridge api or NGRAPH_TF_DISABLE is set
   // we will not do anything; all subsequent
   // passes become a no-op.
-  if (config::IsEnabled() == false ||
-      std::getenv("NGRAPH_TF_DISABLE") != nullptr) {
-    NGRAPH_VLOG(0) << "NGTF_OPTIMIZER: Ngraph is disabled ";
+  bool ngraph_not_enabled =
```
Comment (avijit-nervana) [inline src/grappler/ngraph_optimizer.cc]: Need to replace these repeating code blocks with a utility function.

File: src/ngraph_rewrite_pass.cc [hunk 9]
```diff
@@ -122,8 +123,24 @@ class NGraphVariableCapturePass : public NGraphRewritePass {
     // If ngraph is disabled via ngraph_bridge api or NGRAPH_TF_DISABLE is set
     // we will not do anything; all subsequent
     // passes become a no-op.
-    if (config::IsEnabled() == false ||
-        std::getenv("NGRAPH_TF_DISABLE") != nullptr) {
+    bool ngraph_not_enabled =
+        (!config::IsEnabled()) || (std::getenv("NGRAPH_TF_DISABLE") != nullptr);
```
Comment (avijit-nervana) [inline src/ngraph_rewrite_pass.cc]: May be it's better to absorb the "getenv" stuff inside config::IsEnabled()?

File: src/ngraph_rewrite_pass.cc [hunk 10]
```diff
@@ -122,8 +123,24 @@ class NGraphVariableCapturePass : public NGraphRewritePass {
...
+      // populated. Then we run a new network, it repopulates the cluster
+      // manager. This works under the assumption that whenever
+      // NGraphEncapsulate's Compute is run the rewrite passes (grappler or
+      // optimization passes) have also run (compute --> rewrite). Now that
+      // assumption is broken because now we support NGraphEncapsulate enabled
+      // graphs. Such graphs will not run the first rewrite pass, hence the
+      // cluster manager is not overwritten. Which would mean that cluster
+      // manager contains stale data from a previous run. Hence evicting cluster
+      // manager when rewrite passes are not run.
+      NGRAPH_VLOG(0) << "Not running through nGraph. nGraph not enabled: "
```
Comment (avijit-nervana) [inline src/ngraph_rewrite_pass.cc]: No VLOG(0)

**Pull request merged**